src/core/dev_api/openvino/core/paged_cache_manager.hpp

```diff
@@ -0,0 +1,252 @@
+    void release_blocks(size_t node_id, const std::vector<std::size_t>& blocks);
+
+    // insert and scoring
+    template <typename T>
+    std::vector<std::size_t> insert(size_t node_id,
+                                    const T* key_src,
+                                    const T* value_src,
+                                    std::size_t block_count,
+                                    const T* scores /* may be nullptr */);
+
+    template <typename T>
+    void set_block_scores(size_t node_id, const std::vector<std::size_t>& block_indices, const T* scores);
+
+    // evict to maintain free pool
+    void evict_to_target_free(std::size_t target_free_blocks);
+
+private:
+    inline static size_t m_node_id = 0;
+
+    struct block_t {
+        std::size_t index{0};
+        float score{std::numeric_limits<float>::infinity()};
+        size_t owner{0};
+    };
+
+    struct operator_state {
+        size_t node_id;
+        std::vector<std::size_t> blocks;
+        std::vector<float> scores;
+        std::vector<std::int32_t> subsequence_begins;
+
+        std::size_t num_blocks{0}, num_heads{0}, block_size{0};
+        std::size_t key_head_size{0}, value_head_size{0}, query_head_size{0};
+    };
+
+    // helpers
+    void* offset_key(std::size_t block_idx) const noexcept;
+    void* offset_value(std::size_t block_idx) const noexcept;
+
+    void compute_subsequence_begins_unlocked(operator_state& st) const;
+    void compute_operator_cache_geometry(operator_state& state,
+                                         const size_t block_size,
+                                         const size_t num_heads,
+                                         const size_t key_head_size,
+                                         const size_t value_head_size,
+                                         const size_t query_head_size);
+
+    void ensure_free_blocks_unlocked(std::size_t need_blocks);
+    void evict_one_unlocked();
+
+    std::vector<std::size_t> acquire_blocks_unlocked(size_t node_id, std::size_t block_count);
+    void copy_blocks_into_buffers_unlocked(const void* key_src_bytes,
+                                           const void* value_src_bytes,
+                                           const std::vector<std::size_t>& block_idxs);
+    void set_scores_for_blocks_unlocked(size_t node_id,
+                                        const std::vector<std::size_t>& block_idxs,
+                                        const float* scores);
+
+    static float cast_score_to_float(ov::element::Type et, const void* src_scalar) noexcept;
+    static bool is_element_compatible_with_T(ov::element::Type et, size_t sizeofT) noexcept;
+
+    void rebuild_evict_heap_unlocked();
+    static bool heap_less(const std::vector<block_t>& blocks, std::size_t a, std::size_t b) noexcept;
+
+private:
+    static void* aligned_allocate(std::size_t size);
+    static void aligned_free(void* p);
```

Comment:
Why custom allocation are required.
Ov has AlignedBuffer or standard containers pointers can be use to allocate memory